### PR TITLE
Splitting client and daemon code

### PR DESCRIPTION
**Pull request by FriedrichAltheide**
- Splitted proxy code to a lib containing the client and a daemon side as well as a separate crate containing a proxy
- Moved all dependency definitions to the virtual manifest

### CODE REVIEW

File: nix-remote/src/nix_client.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    sync::Arc,
+};
+
+use crate::{NixRead, NixWrite, Result, PROTOCOL_VERSION, WORKER_MAGIC_1, WORKER_MAGIC_2};
+
+pub struct NixDaemonClient<R, W> {
+    rx_from_daemon: NixRead<R>,
+    tx_to_daemon: NixWrite<W>,
+    tx_op_count: u64,
+    last_op: Option<Arc<WorkerOp>>,
```
Comment (jneem) [inline nix-remote/src/nix_client.rs:22]: Why do we need to keep this?

File: nix-remote/src/nix_daemon_proxy.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    last_op: Option<Arc<WorkerOp>>,
+}
+
+impl<R: Read, W: Write> NixDaemonProxy<R, W> {
+    pub fn new(r: R, w: W) -> Result<Self> {
+        let mut daemon = Self {
+            rx_from_client: NixRead { inner: r },
+            tx_to_client: NixWrite { inner: w },
+            rx_op_count: 0,
+            last_op: None,
```
Comment (jneem) [inline nix-remote/src/nix_daemon_proxy.rs:25]: Why do we need to keep this?

**Review by jneem — COMMENTED** [review]
Thanks for the PR! This looks good in isolation, but it's also going to conflict with #5 and #8. @YorikSar does this re-org meet what you need?

File: nix-remote/src/nix_client.rs (hunk 1 above)
Comment (FriedrichAltheide) [reply nix-remote/src/nix_client.rs]: It is now removed

File: nix-remote/src/nix_daemon_proxy.rs (hunk 2 above)
Comment (FriedrichAltheide) [reply nix-remote/src/nix_daemon_proxy.rs]: It is now removed

**Comment by YorikSar** [comment]
@jneem This looks much better than my attempts at it. I also won't have time to focus on this any time soon. I'll close my PRs in favour of this one.

**Review by jneem — APPROVED** [review]

**Pull request merged**